wrapanapi/systems/google.py

```diff
@@ -0,0 +1,798 @@
+
+            Args:
+                project: name of the project, so called project_id
+                zone: zone of cloud
+                service_account: service_account_content
+
+                scope: compute engine, container engine, sqlservice end etc
+                file_path: path to json or p12 file
+                file_type: p12 or json
+                client_email: Require for p12 file
+
+            Returns: A :py:class:`GoogleCloudSystem` object.
+        """
+        super(GoogleCloudSystem, self).__init__(**kwargs)
+        self._project = project
+        self._zone = zone
+        self._region = kwargs.get('region')
+        scope = kwargs.get('scope', self.default_scope)
+
+        service_account = kwargs.get('service_account', None)
+        if service_account:
+            service_account = dict(service_account.items())
+            service_account['private_key'] = service_account['private_key'].replace('\\n', '\n')
+            credentials = ServiceAccountCredentials.from_json_keyfile_dict(
+                service_account, scopes=scope)
+        elif file_type == 'json':
+            file_path = kwargs.get('file_path', None)
+            credentials = ServiceAccountCredentials.from_json_keyfile_name(
+                file_path, scopes=scope)
+        elif file_type == 'p12':
+            file_path = kwargs.get('file_path', None)
+            client_email = kwargs.get('client_email', None)
+            credentials = ServiceAccountCredentials.from_p12_keyfile(
+                client_email, file_path, scopes=scope)
+        http_auth = credentials.authorize(httplib2.Http())
+        self._compute = build('compute', 'v1', http=http_auth)
+        self._storage = build('storage', 'v1', http=http_auth)
+        self._instances = self._compute.instances()
+        self._forwarding_rules = self._compute.forwardingRules()
+        self._buckets = self._storage.buckets()
+
+    @property
+    def _identifying_attrs(self):
+        return {'project': self._project, 'zone': self._zone, 'region': self._region}
+
+    def _get_all_buckets(self):
+        return self._buckets.list(project=self._project).execute()
+
+    def _get_all_forwarding_rules(self):
+        results = []
+        results.extend(self._forwarding_rules.list(project=self._project, region=self._zone).
+                       execute().get('items', []))
+        return results
+
+    def info(self):
+        return "{}: project={}, zone={}".format(self.__class__.__name__, self._project, self._zone)
+
+    def disconnect(self):
+        """
+        Disconnect from the GCE
+
+        GCE service is stateless, so there's nothing to disconnect from
+        """
+        pass
+
+    def list_vms(self, zones=None):
+        """
+        List all VMs in the GCE account, filtered by zone if desired
+
+        Args:
+            zone: List of zones, by default does not filter to any zone
```

Comment:
So, this being plural is a departure from the current implementation of singular method names.  Its a good change - but its one to be aware of the image of.

Second, this behaves distinctly differently from the existing behavior where `list_vm()`  filters by zone, by default. The `all_vms` method provides this behavior through an argument, and it by default returns an unfiltered list of vms.

I think that the current behavior should be maintained in this case, where my zone-restricted mgmt object's `list_vms` function only returns vm's in the already restricted zone.

Reply:
Changing to use `self._zone` by default